Checkpoint defense feedback

warenga opened the pull request:
#### What does this PR do?
- Adds IndexInstance class functionality to save files uploaded.
- Changes createIndex front end functionality to be called when user clicks 'Create Index' button instead of when a file is uploaded.
- Modifies test suits to test via the new changes made in InvertedIndex class.
- Adds browser-sync and gulp-nodemon to gulpfile to better reload browser and server when public files change .
- Adds functionality in isJSON method to check if JSON file has a title and a text. 

 #### Description of Task to be completed?
To implement the following feedback received from checkpoint defense 2:
<img width="707" alt="screen shot 2017-01-27 at 4 19 56 pm" src="https://cloud.githubusercontent.com/assets/16874615/22372223/7a5ef31e-e4ac-11e6-974a-6b92b25a316f.png">

Commit:
Add test suits under Populate Index

Commit:
Add catch invalid JSON error

Commit:
Add reverse.json file

Commit:
fix [object Object] test bug

Commit:
add arrow function, remove unnecessary code

Commit:
Add coverage badge

Commit:
(chore): change InvertedIndex class functionality

Add files object to save all files.
Change methods to accept filename parameters.
Change isJson to check if json has title and text field.

Commit:
(feature): change create index functionality

Removed createIndex being called from fileModel to the MainController. This allows createIndex to be called when the button 'Create Index' is clicked.

Commit:
(feature): Add browser-sync and gulp-nodemon

Add browser-sync and gulp-nodemon to reload browser and server when files in public folder change.

Commit:
browser-sync and gulp-nodemon added to dependencies

Commit:
passed appropriate parameters to indexInstance method calls

Commit:
remove secondInstance, fixed Populate index override test bug

Commit:
(chore): eslint inverted-index

Add semi-colon where necessary.

Commit:
(chore): properly format docblock

Commit:
(chore): Add comments for better documentation

Commit:
(chore): Add more tests for better coverage

Commit:
(chore): feedback file modified

Commit:
(fix): format code to standard specified

andela-kakpobome commented on jasmine/spec/inverted-index-test_spec.js:
```diff
@@ -23,20 +24,38 @@ describe('Read book data', () => {
   });
+
+  it('should have a tile and text field', () => {
+    index.createIndex('invalid2.json', invalid)
```
### Critical 

@andela-wmaina you might want to listen to HoundCI though, please fix!

andela-kakpobome commented on server.js:
```diff
@@ -1,36 +1,36 @@
+ });
 
-app.listen(1337, () => {});
+ app.listen(1337, () => {});
```
### Nitpick

The anonymous function serves no purpose here. We can listen by passing in just the port number.

andela-kakpobome commented on jasmine/spec/inverted-index-test_spec.js:
```diff
@@ -23,20 +24,38 @@ describe('Read book data', () => {
   });
+
+  it('should have a tile and text field', () => {
+    index.createIndex('invalid2.json', invalid)
```
### Critical

You might want to listen to HoundCI here!

andela-kakpobome reviewed (changes requested):
Please attend to the issues I raised so we can close this

Commit:
(fix): add semicolon to line 29

Commit:
(fix): remove unnecessary anonymous function

Commit:
(fix): add semicolon to line 30

andela-kakpobome commented on jasmine/spec/inverted-index-test_spec.js:
```diff
@@ -48,20 +67,39 @@ describe('Search index', () => {
+    index.createIndex('test.json', test);
+    const results = [];
+    Object.keys(index.files).forEach((obj) => {
+      results.push(index.searchIndex(obj, 'alliance', 'alice', 'powerful', 'comical'));
```
### Nitpick

`obj` should actually be filename, because that's exactly what it is, your naming is confusing, please rectify.

andela-kakpobome commented on public/js/controllers/MainController.js:
```diff
@@ -1,64 +1,88 @@
+      * createIndex
+      *
+      * sends a http.post to server with the file.
+      * @params {string}
```
### Nitpick

There should have been a line space before the method description.

andela-kakpobome commented on public/js/controllers/MainController.js:
```diff
@@ -1,64 +1,88 @@
+      * passes the selected files and the searched words 
+      * to search function
+      *
+      * @param none
```
### Nitpick

Even though it's not explicitly defined in the method signature, it's not entirely accurate that the `search` does not accept any arguments.

andela-kakpobome commented on public/js/directives/fileModel.js:
```diff
@@ -4,34 +4,53 @@ app.directive('fileList', ['$http', '$localStorage', function($http, $localStora
+          const reader = new FileReader();
+
           // inject file on onload
           reader.onload = function(event) {
```
### Critical

This should have been a big arrow function

andela-kakpobome commented on public/js/directives/fileModel.js:
```diff
@@ -4,34 +4,53 @@ app.directive('fileList', ['$http', '$localStorage', function($http, $localStora
+              // Checks if file is a valid JSON file
+              if (isJSON(fileContent)) {
+                scope.files.push(evt.target.files[0].name);
+                let fileName = evt.target.files[0].name;
```
### Critical

This should have been a `const`

andela-kakpobome commented on public/js/directives/fileModel.js:
```diff
@@ -4,34 +4,53 @@ app.directive('fileList', ['$http', '$localStorage', function($http, $localStora
-        }).error(function(data) {
-          alert('error');
-        });
+      let isJSON = (fileContent) => {
```
### Critical

@andela-wmaina Is there any value to having this method duplicated both on the front and server side?

warenga replied:
You are right. I have replaced this with a http post that sends the fileContent to the server and the server calls the isJSON method.

Commit:
(chore): changed variable name 'obj' to 'filename'

Commit:
(chore): specified the params needed for the searchIndex method

Commit:
(chore): removed isJSON function

removes the repeated isJSON function and added a http.post to check the validity of the file.

Commit:
(chore): fixed eslint errors

Commit:
(chore): changed isJSON function to accept an object

Commit:
(fix): removed unnecessary space

andela-kakpobome commented on src/inverted-index.js:
```diff
@@ -1,85 +1,126 @@
+ }
 
-exports.InvertedIndex = InvertedIndex;
+ exports.InvertedIndex = InvertedIndex;
```
### Critical

@andela-wmaina Use the es6 specified approach for exporting classes. i.e `export class` and leverage Babel if you need to.

andela-kakpobome commented on jasmine/files/test.json:
```diff
@@ -0,0 +1,10 @@
+[{
```
### Nitpick

Please ensure you use the following format when declaring multiple objects in a collection:
```
[
   {
      x: 'value'
   },
   {
      y: 'value'
   }
]
```
